test-resources/headless-core-stub/lambdas/start/src/services/signing-service.ts

```diff
@@ -1,16 +1,82 @@
 import { GetPublicKeyCommand, KMSClient } from "@aws-sdk/client-kms";
-import { CompactEncrypt, importSPKI, KeyLike } from "jose";
+import { CompactEncrypt, importJWK, importSPKI, JWK, KeyLike } from "jose";
+import {
+    clearJWKSCache,
+    fetchAndCacheJWKS,
+    getCachedJWKS,
+    isJWKSCacheValid,
+} from "../../../../utils/src/jwks-cache-control";
 import { HeadlessCoreStubError } from "../../../../utils/src/errors/headless-core-stub-error";
+import { formatAudience } from "../../../../utils/src/audience-formatter";
+import { Logger } from "@aws-lambda-powertools/logger";
 
 const kmsClient = new KMSClient({ region: "eu-west-2" });
+export const logger = new Logger();
 
 let cachedPublicKey: KeyLike | undefined;
 
-export const getPublicEncryptionKey = async () => {
-    if (cachedPublicKey) {
+export const getPublicEncryptionKey = async (audience: string) => {
+    if (isJWKSCacheValid()) {
         return cachedPublicKey;
     }
 
+    if (process.env.KEY_ROTATION_FEATURE_FLAG_ENABLED === "true") {
+        await setCachedPublicEncryptionKeyFromJwks(audience);
+    }
+
+    if (!getCachedJWKS()) {
+        logger.info({ message: "using KMS to retrieve encryption key" });
+        await getPublicEncryptionKeyFromKms();
+    }
+
+    return cachedPublicKey;
+};
+
+export const encryptSignedJwt = (signedJwt: string, publicEncryptionKey: KeyLike) => {
+    return new CompactEncrypt(new TextEncoder().encode(signedJwt))
+        .setProtectedHeader({ alg: "RSA-OAEP-256", enc: "A256GCM" })
+        .encrypt(publicEncryptionKey);
+};
+
+export function _resetCachedPublicKeyForTest() {
+    clearJWKSCache();
+}
+
+async function setCachedPublicEncryptionKeyFromJwks(audience: string) {
+    const audienceApi = formatAudience(audience);
+    const criEncryptionJwksEndpoint = new URL("/.well-known/jwks.json", audienceApi);
+
+    logger.info({ message: "Attempting to use CRI hosted public encryption jwks endpoint", criEncryptionJwksEndpoint });
+
+    if (isJWKSCacheValid()) {
+        logger.info({ message: "Using locally cached JWKs", criEncryptionJwksEndpoint, ...getCachedJWKS() });
+    } else {
+        logger.info({ message: "JWKS cache expired or missing; fetching new JWKS..." });
+        await fetchAndCacheJWKS(criEncryptionJwksEndpoint, logger);
+    }
+
+    const jwks = getCachedJWKS();
+
+    if (!jwks?.keys || !Array.isArray(jwks?.keys) || jwks?.keys?.length === 0) {
+        clearJWKSCache();
+        logger.info({ message: "Invalid or empty JWKS response", criEncryptionJwksEndpoint });
+    }
+    logger.info({ message: "Successfully retrieved public encryption jwks endpoint", ...jwks });
+
+    const encryptionKey = jwks?.keys
+        .slice()
+        .reverse()
+        .find((k) => k.use === "enc") as JWK;
+
+    if (!encryptionKey) {
+        logger.info(`No encryption key (use: "enc") found in JWKS from ${criEncryptionJwksEndpoint}`);
```

Comment:
This is a nice, helpful message!

Same comment as above, should it be an error log and should we just `return` here?

Reply:
The idea is it fall back eventually to use the old approach which is really on KMS and not the endpoint